Add GDS Session documentation (#587)
neo4j/graph-data-science-client: doc/modules/ROOT/pages/appendix/gds-session.adoc

Available sizes are listed <<gds-session-sizes, below>>.

- **DBMS connection**.
This is a `DbmsConnectionInfo` object that contains the URI of an AuraDB instance, a username, and a password.

.Creating a GDS Session:
[source, python, role=no-test]
----
from graphdatascience.session import SessionSizes, DbmsConnectionInfo

name = "my-new-session"
size = SessionSizes.by_memory().DEFAULT
db_connection_info = DbmsConnectionInfo("neo4j+s://mydbid.databases.neo4j.io", "my-user", "my-password")

gds = sessions.get_or_create(
    session_name=name,
    size=size,
    db_connection=db_connection_info,
)
----

The `get_or_create()` method creates a new GDS Session if it does not exist, or connects to an existing one if it does.
If the session size is different from the existing one, an error is raised.

The return value of `get_or_create()` is an `AuraGraphDataScience` object.
It offers a similar API to the `GraphDataScience` object, but it is configured to run on a GDS Session.
As a convention, always use the variable name `gds` for the return value of `get_or_create()`.

[[gds-session-sizes]]
==== Session sizes

Session sizes are specified using the `SessionSizes.by_memory()` enum.

> Possible values are `DEFAULT`, `XS`, `S`, `SM`, `M`, `ML`, `L`, `XL`, `XXL`, `X3L`, `X4L`, `X5L`.

Reviewer: ```suggestion
Possible values are `DEFAULT` (same as `M`), `XS`, `S`, `SM`, `M`, `ML`, `L`, `XL`, `XXL`, `X3L`, `X4L`, `X5L`.
```

)
----

With the `gds` GDS Session active, project the graph and specify node and relationship property schemas as follows:

[source,python,role=no-test]
----
from graphdatascience.session import GdsPropertyTypes

G, result = gds.graph.project(
    "my-graph",
    """
    CALL {
        MATCH (u1:User)
        OPTIONAL MATCH (u1)-[r:KNOWS]->(u2:User)
        RETURN u1 AS source, r AS rel, u2 AS target, {} AS sourceNodeProperties, {} AS targetNodeProperties
        UNION
        MATCH (p:Product)
        OPTIONAL MATCH (p)<-[r:BOUGHT]-(user:User)
        RETURN user AS source, r AS rel, p AS target, {} AS sourceNodeProperties, {cost: p.cost} AS targetNodeProperties
    }
    RETURN gds.graph.project.remote(source, target, {
      sourceNodeProperties: sourceNodeProperties,
      targetNodeProperties: targetNodeProperties,
      sourceNodeLabels: labels(source),
      targetNodeLabels: labels(target),
      relationshipType: type(rel),
      relationshipProperties: properties(rel)
    })
    """,
    nodePropertySchema={"cost": GdsPropertyTypes.DOUBLE},
    relationshipPropertySchema={"since": GdsPropertyTypes.LONG, "price": GdsPropertyTypes.DOUBLE},
)
----

== Running algorithms

> You can run algorithms on a remotely projected graph in the same way as you would on any other graph.
> For a full list of available algorithms, see the https://neo4j.com/docs/graph-data-science-client/{page-version}/api/[API reference].
> 
> === Example 
> 
> Extending the previous example, we will now run the PageRank and FastRP algorithms on the projected graph.
> Then we will stream the results back joined with the `name` property fetched from the db.

Reviewer: ```suggestion
You can run algorithms on a remotely projected graph in the same way you would on any projected graph.
For instance, you can run the PageRank and FastRP algorithms on the projected graph from the previous example as follows:
```

Author: So you are recommending to not have extra `Example` subheadings for each section?

Author: yeah let's skip the additional subheadings

Reviewer: Yes, just because there's not much under the heading, nor it is a separate example.

[source,python,role=no-test]
----
gds.pageRank.mutate(G, mutateProperty="pr")
gds.fastRP.mutate(G, featureProperties=["pr"], embeddingDimension=2, nodeSelfInfluence=0.1, mutateProperty="embedding")

> gds.graph.nodeProperties.stream(G, db_properties=["name"], node_properties=["pr", "embedding"])

Reviewer: ```suggestion
# Stream the results back together with the `name` property fetched from the database
gds.graph.nodeProperties.stream(G, db_properties=["name"], node_properties=["pr", "embedding"])
```

Also, is db_properties correct? Shouldn't it be db_node_properties?

----

> 

Reviewer: ```suggestion

For a full list of the available algorithms, see the https://neo4j.com/docs/graph-data-science-client/{page-version}/api/[API reference].
```

== Remote write-back

> The GDS Session's in-memory graph was projected from data in AuraDB.
> Write back operations will thus persist the data back to the same AuraDB.
> When calling any write operations the GDS Python client will automatically use the remote write back functionality.
> This includes all `.write` algorithm modes as well as all `.write` graph operations.

Reviewer: ```suggestion
The GDS Session's in-memory graph was projected from data in AuraDB, so write-back operations will persist the data back to the same AuraDB instance.
When calling any write operations, the GDS Python client will automatically use the remote write-back functionality.
This includes all `.write` algorithm modes as well as all `.write` graph operations.
```

> === Example
> 
> Extending the previous example, we will now write back the FastRP embeddings to the AuraDB instance.

Reviewer: ```suggestion

Extending the previous example, you can write back the FastRP embeddings to the AuraDB instance as follows:
```

[source,python,role=no-test]
----
gds.graph.nodeProperties.write(G, "embedding")
----

== Querying the database

You can run Cypher queries on the AuraDB instance using the `run_cypher()` method.
There is no restriction on the type of query that can be run, but it is important to note that the query will be run on the AuraDB instance, and not on the GDS Session.

> Thus you will not be able to call gds operations from the `run_cypher()` method.

Reviewer: ```suggestion
Therefore, you will not be able to call any GDS procedures from the `run_cypher()` method.
```

[source,python,role=no-test]
----
gds.run_cypher("MATCH (n:User) RETURN n.name, n.embedding")
----

> We have now integrated our analytical results with the original data in the AuraDB instance.
> 

Reviewer: ```suggestion
```

I think we can skip this too.

A GDS Session is a temporary compute environment for running GDS workloads.
It is a service offered by Neo4j and runs within Neo4j's Aura cloud platform.
A GDS Session reads data from a Neo4j DBMS system through a _remote projection_, then runs computations on the projected graph, and lastly can write back the analytical results using _remote write-back_.

NOTE: GDS Sessions are only available for Neo4j AuraDB Enterprise, and not by default. Contact your account manager to get the features enabled.

== GDS Session management

The `GdsSessions` object is the top-level API which allows you three operations:

- `get_or_create`: Create a new GDS Session, or connect to an existing one.
- `list`: List all currently active GDS Sessions.
- `delete`: Delete a GDS Session.

=== Pre-requisites

To create a GDS Session, you need to have a Neo4j Aura account.
Using your account, you need to create Aura API credentials.
These credentials are a pair of `CLIENT_ID` and `CLIENT_SECRET` that you will use to authenticate with the Aura API.
For how to create these credentials, see the https://neo4j.com/docs/aura/platform/api/authentication/#_creating_credentials[Aura documentation].

> Using these credentials, you will create a `GdsSessions` object, and use it to create a new GDS Session.

Author: very succinct!

.Creating a GdsSessions object:
[source, python, role=no-test]
----
from graphdatascience.session import GdsSessions, AuraAPICredentials

CLIENT_ID = "my-aura-api-client-id"
CLIENT_SECRET = "my-aura-api-client-secret"

# Create a new GdsSessions object
sessions = GdsSessions(ds_connection=AuraAPICredentials(CLIENT_ID, CLIENT_SECRET))
----

=== Creating a GDS Session

To create a GDS Session, you need to provide the following information:

- **Session name**.
The name must be unique.

- **Session size**.
The size determines the amount of memory and CPU available to the session.
It also determines the cost of running the session.
Available sizes are listed in xref:gds-session-sizes[].

- **DBMS connection**.
This is a `DbmsConnectionInfo` object that contains the URI of an AuraDB instance, a username, and a password.

> With these pieces of information, we are ready to create a GDS Session.

Author: Is this just superfluous? Like too talkative?

.Creating a GDS Session:
[source, python, role=no-test]
----
from graphdatascience.session import SessionSizes, DbmsConnectionInfo

name = "my-new-session"
size = SessionSizes.by_memory().DEFAULT
db_connection_info = DbmsConnectionInfo("neo4j+s://mydbid.databases.neo4j.io", "my-user", "my-password")

gds = sessions.get_or_create(
    session_name=name,
    size=size,
    db_connection=db_connection_info,
)
----

The `get_or_create()` method will create a new GDS Session if it does not exist, or connect to an existing one if it does.

> If the session size is different from the existing one, an error will be raised.

Author: Is the docs convention to use present form?

The return value of `get_or_create()` is an `AuraGraphDataScience` object.
It offers a similar API to the `GraphDataScience` object, but it is configured to run on a GDS Session.
By convention, always use the variable name `gds` for the return value of `get_or_create()`.

[[gds-session-sizes]]
==== Session sizes

Session sizes are logical and specified using the `SessionSizes.by_memory()` enum.

> Possible values are `DEFAULT`, `XS`, `S`, `SM`, `M`, `ML`, `L`, `XL`, `XXL`, `X3L`, `X4L`, `X5L`.

Author: My ambition here is to map the logical size to a monetary cost, but not necessarily to a physical size. So I'm endeavouring to not expose the actual memory used. Now that is perhaps a bit strict; we may as well have a mapping layer that goes from "physical size" to actual physical size.

But to your direct Qs:

- DEFAULT = M
- No

> Possible values are `DEFAULT`, `XS`, `S`, `SM`, `M`, `ML`, `L`, `XL`, `XXL`, `X3L`, `X4L`, `X5L`.

Reviewer: I see. Yet, there is a mention of a `by_memory` enum, which would lead me to think they are still somewhat related to physical memory.

How can we explain a little better what they mean without being too specific? If we don't explain, a user would have no way to choose the appropriate one.

> Possible values are `DEFAULT`, `XS`, `S`, `SM`, `M`, `ML`, `L`, `XL`, `XXL`, `X3L`, `X4L`, `X5L`.

Author: Yeah there was a requirement for having two dimensions that sessions would be sized after: a CPU-scaling one and a Memory-scaling one. So far we only have implemented one of the size dimensions, but didn't want to occupy the namespace to block us in the future, hence `by_memory`. Without the other dimension it doesn't look good, I agree with that. I will note that down and see if we can come up with something better.

We've asked explicitly about feedback on the sizing API but received only a little. It is indeed related to physical memory, but my ambition is to not work with concrete numbers of bytes, rather preferring relative measures.

We are working on exposing a size estimator tool, which would recommend a particular size based on a user's provided node/rel counts and maybe selected algorithms (think AuraDS console; same underlying software). 

I think in principle, it is often difficult to know how large of a cloud instance you need upfront. But switching should be easy when you get it too small. I would argue that applies also here. Without the above tool, I could think of adding a part here recommending something like

>Start out with a smaller instance. If it is too small, delete it and create a larger one.

It's pretty raw. Do you think it is better than nothing?

With the tool available, I think of something like

>Use the tool to get an initial size recommendation. If it turns out to be too small, delete it and create a larger one.

The idea here is that since the GDS Session doesn't represent any persistent state, deleting and starting over is an OK thing to do. This mirrors my own experience of using e.g. AWS EC2.

> Possible values are `DEFAULT`, `XS`, `S`, `SM`, `M`, `ML`, `L`, `XL`, `XXL`, `X3L`, `X4L`, `X5L`.

Reviewer: Ah, I see. Then I agree, let's wait for the estimator tool so we have something useful to say.

|===

==== Remote projection query syntax

The remote projection query supports the same syntax as a Cypher projection, with two key differences:

. The graph name is not a parameter.
Instead, the graph name is provided to the `gds.graph.project()` endpoint.
. The `gds.graph.project.remote()` function must be used, instead of the `gds.graph.project()` function.

For full details and examples on how to write Cypher projection queries, see the https://neo4j.com/docs/graph-data-science/current/management-ops/graph-creation/graph-project-cypher-projection/[Cypher projection documentation in the GDS Manual].

==== Property schemas

The optional parameters `nodePropertySchema` and `relationshipPropertySchema` are useful when projecting graphs with multiple node labels or relationship types which have distinct property sets.
We call such graphs _heterogeneous graphs_.
If these parameters are not provided, the type of each property will be inferred from the first row of data seen by the projection function.
For homogeneous graphs, this is usually sufficient, but for heterogeneous graphs, it becomes cumbersome to provide default values for each property on every row of data.

> We recommend to use the property schema parameters always.

Author: No it is not just about performance, although it is about that too. The additional thing is that without the schema parameters, a user would have to project placeholder values for properties that are missing for a label. For example:

```
CALL {
  MATCH (p1:Person)-->(p2:Person)
  RETURN p1 AS source, p2 AS target, p1 { .age, cost: -1 } AS sp
  UNION
  MATCH (p1:Product)-->(p2:Product)
  RETURN p1 AS source, p2 AS target, p1 { .cost, age: -1 } AS sp
}
RETURN gds.graph.project(source, target, {sourceNodeProperties: sp})
```

Here we have to specify both `age` and `cost` even though we have only one of them present in each arm of the `UNION`. This is a bit complicated to explain in the documentation though, and it is easier to recommend to always use the schemas. But when there is only one node type there is no problem.

> If these parameters are not provided, the type of each property will be inferred from the first row of data seen by the projection function.
> For homogeneous graphs, this is usually sufficient, but for heterogeneous graphs, it becomes cumbersome to provide default values for each property on every row of data.
> We recommend to use the property schema parameters always.

Reviewer: ```suggestion

If these parameters are not provided, the type of each property is inferred from the first row of data seen by the projection function.
For homogeneous graphs this is usually sufficient, while for heterogeneous graphs it is not.
For this reason, we recommend to always use the property schema parameters.
```

I see, thanks for the explanation. How about this instead?

> We recommend to use the property schema parameters always.

Author: Reads great to me, thank you.

==== Relationship type undirectedness and inverse indexing

The optional parameters `undirectedRelationshipTypes` and `inverseIndexedRelationshipTypes` are used to configure undirectedness and inverse indexing of relationships.
These have the same behaviour as is documented in the https://neo4j.com/docs/graph-data-science/current/management-ops/graph-creation/graph-project-cypher-projection/#graph-project-cypher-projection-syntax[GDS Manual].

=== Example

In this example, we will illustrate how to project a graph into a GDS Session.
The example graph is heterogeneous and models users and products.
Users can know each other, and users can buy products.

> We assume here that the database connection is to a new, empty AuraDB instance.

Author: In this example the user must bring their own db. This way of stating it also makes our expectation clear.

> .First, we create a GDS Session and add some data to our database.

Author: Do you think we are good with no heading at all for the code block?

> .First, we create a GDS Session and add some data to our database.

Reviewer: I saw that in the block right below there is no heading too, so I thought we can skip it. If you think it's better to keep it, I'll just slightly rephrase to:

`Create a GDS Session and add some data to the database.`

> .First, we create a GDS Session and add some data to our database.

Author: I like having a heading for all code blocks, as a principle. If you think I'm too rigid please tell :)

> .First, we create a GDS Session and add some data to our database.

Reviewer: Nothing against it, as long as it provides useful information :)

[source,python,role=no-test]
----
import os # for reading environment variables
from graphdatascience.session import SessionSizes, DbmsConnectionInfo, GdsSessions, AuraAPICredentials

sessions = GdsSessions(ds_connection=AuraAPICredentials(os.environ["CLIENT_ID"], os.environ["CLIENT_SECRET"]))

db_connection = DbmsConnectionInfo(os.environ["DB_URI"], os.environ["DB_USER"], os.environ["DB_PASSWORD"])
gds = sessions.get_or_create(
    session_name="my-new-session",
    size=SessionSizes.by_memory().DEFAULT,
    db_connection=db_connection,
)

gds.run_cypher(
    """
    CREATE
     (u1:User {name: 'Mats'}),
     (u2:User {name: 'Florentin'}),
     (p1:Product {name: 'ice cream', cost: 4.2}),
     (p2:Product {name: 'computer', cost: 13.37})

    CREATE
     (u1)-[:KNOWS {since: 2020}]->(u2),
     (u2)-[:BOUGHT {price: 7474}]->(p1),
     (u1)-[:BOUGHT {price: 1337}]->(p2)
    """
)
----

> Next, we project the graph into the GDS Session.
> There are some node and relationship properties, which we specify using the `nodePropertySchema` and `relationshipPropertySchema` parameters.

Reviewer: ```suggestion
With the `gds` GDS Session active, project the graph and specify node and relationship property schemas as follows:
```

[source,python,role=no-test]
----
from graphdatascience.session import GdsPropertyTypes

G, result = gds.graph.project(
    "my-graph",
    """
    CALL {
        MATCH (u1:User)
        OPTIONAL MATCH (u1)-[r:KNOWS]->(u2:User)
        RETURN u1 AS source, r AS rel, u2 AS target, {} AS sourceNodeProperties, {} AS targetNodeProperties
        UNION
        MATCH (p:Product)
        OPTIONAL MATCH (p)<-[r:BOUGHT]-(user:User)
        RETURN user AS source, r AS rel, p AS target, {} AS sourceNodeProperties, {cost: p.cost} AS targetNodeProperties
    }
    RETURN gds.graph.project.remote(source, target, {
      sourceNodeProperties: sourceNodeProperties,
      targetNodeProperties: targetNodeProperties,
      sourceNodeLabels: labels(source),
      targetNodeLabels: labels(target),
      relationshipType: type(rel),
      relationshipProperties: properties(rel)
    })
    """,
    nodePropertySchema={"cost": GdsPropertyTypes.DOUBLE},
    relationshipPropertySchema={"since": GdsPropertyTypes.LONG, "price": GdsPropertyTypes.DOUBLE},
)
----

> We have now successfully projected a graph from the AuraDB instance to the new GDS Session and are ready to run algorithms.

Reviewer: ```suggestion
```

Author: I've developed a habit of embedding code blocks with paragraphs, but perhaps it isn't always necessary.

Reviewer: I think that it may depend on the length. If the code block is long or needs additional explanation, by all means. In this case IMHO this specific line just repeats what you set to do.
